ignite/distributed/comp_models/base.py

```diff
@@ -256,18 +261,18 @@ def create_from_context() -> "_SerialModel":
         return _SerialModel()
 
     @staticmethod
-    def create_from_backend(backend: Optional[str] = None, **kwargs) -> "_SerialModel":
+    def create_from_backend(backend: Optional[str] = None, **kwargs: Any) -> "_SerialModel":
         return _SerialModel()
 
     @staticmethod
-    def spawn(*args, **kwargs):
+    def spawn(*args: Any, **kwargs: Any) -> None:
         raise NotImplementedError("Serial computation model does not implement spawn method")
 
     def all_reduce(self, tensor: Union[torch.Tensor, Number], op: str = "sum") -> Union[torch.Tensor, Number]:
         return tensor
 
-    def all_gather(self, tensor: Union[torch.Tensor, Number]) -> Union[torch.Tensor, Number]:
-        return tensor
+    def all_gather(self, tensor: Union[torch.Tensor, Number, str]) -> Union[torch.Tensor, Number, List[str]]:
```

Comment:
Is it correct `Union[torch.Tensor, Number, str]` and `Union[torch.Tensor, Number, List[str]]` ?
Seems like they are not consistent with below `cast(Union[torch.Tensor, Number], tensor)`. Can we keep the typing without `str` ?

Reply:
if I do so, several issues will pop up
```
ignite/distributed/comp_models/base.py:274: error: Argument 1 of "all_gather" is incompatible with supertype "ComputationModel"; supertype defines the argument type as "Union[Tensor, Number, str]"  [override]
```
this I can fix with `# type: ignore`, but then I also get
```
ignite/distributed/utils.py:352: error: Argument 1 to "all_gather" of "_SerialModel" has incompatible type "Union[Tensor, Number, str]"; expected "Union[Tensor, Number]"  [arg-type]
```
because of the return type of the Function
https://github.com/pytorch/ignite/blob/221b6dedd95ef6187ba1989013dfc49049ebcb62/ignite/distributed/utils.py#L332-L347
but of course, I can also put an `# type: ignore` there 😄

Reply:
OK, it makes sense to implement it in the same way as distributed implementations. Let's do this in another PR.
Idea is to rework types for a) `utils.all_gather` and all its realizations: serial, native, hvd, xla. 
Thinking more about its type annotations, i'd say it should be 
```python
def all_gather(tensor: Union[torch.Tensor, Number, str]) -> Union[torch.Tensor, List[Number], List[str]]:
```
Such that here in _SerialModel it should be implemented as 
```python
def all_gather(self, tensor: Union[torch.Tensor, Number, str]) -> Union[torch.Tensor, List[Number], List[str]]:
    if isinstance(tensor, torch.Tensor):
        return tensor
    return cast(Union[List[Number], List[str]], tensor)
```

Reply:
ok, I will add the `# type: ignore` and it should be fixed via a different PR as you suggested 👍